fix: avro_to_arrow: Handle avro nested nullable struct (union)

Samrose-Ahmed opened the pull request:
Corrects handling of a nullable struct union.

## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #7662.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->
Bug fix

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

alamb commented:
@Samrose-Ahmed  -- is it possible to add a test to ensure this functionality is not broken in the future?

alamb commented:
(Thank you for the contribution, BTW)

Samrose-Ahmed commented:
Of course I can add a test.

Samrose-Ahmed commented:
I have added a test to verify this behavior.

alamb commented:
I don't think the CI failures are related to your work

alamb reviewed (approved):
Looks good to me -- thank you @Samrose-Ahmed

alamb commented:
I took the liberty of merging up from main to get the CI to pass cleanly.

sarutak commented:
@Samrose-Ahmed Thank you for catching this issue! I should have fixed this in #7525 .
@alamb Do we need end-to-end test for this change?

alamb commented:
> @alamb Do we need end-to-end test for this change?

Hi @sarutak  -- end to end test coverage would definitely be better ❤️  --  I am not familiar enough with the avro spec to know how hard/easy that would be to create. Is it something you can help with?

sarutak commented:
>  Is it something you can help with?

@alamb 
Of course! I'll update `nested_record.avro` in `arow-testing` to add nullable record schema.

@Samrose-Ahmed 
After I have finished the change, could you add end-to-end test to `avro.slt`?

Samrose-Ahmed commented:
Sure I can do that

sarutak commented:
@Samrose-Ahmed Or, do you want to change the test data too?
If you want to do it, [this PR](https://github.com/apache/arrow-testing/pull/91) might help you.

Samrose-Ahmed commented:
I would appreciate if you can do it.

sarutak commented:
@Samrose-Ahmed 
OK, let me do it.

sarutak commented:
I've opened a PR to update the test data.
https://github.com/apache/arrow-testing/pull/94

sarutak commented on datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:
```diff
@@ -775,14 +775,20 @@ impl<'a, R: Read> AvroArrowArrayReader<'a, R> {
                             .enumerate()
                             .map(|(i, row)| (i, self.field_lookup(field.name(), row)))
                             .map(|(i, v)| {
+                                let v = v.map(maybe_resolve_union);
```
I think we need to care about array of union of `null` and records as well.

alamb commented:
https://github.com/apache/arrow-testing/pull/94 is merged

Samrose-Ahmed commented:
I've updated the PR. 

I've fixed a correctness issue needed for this PR with mapping the Avro and Arrow schema as well, the Avro Record name should not be used in Arrow (the code that was doing `&format!("{}.{}", name.fullname(None), field.name)`, only the field names should be used, the record name is an Avro concept used for schema merging/references. I've added a test for that as well.

Samrose-Ahmed commented:
Actually this code is a bit weird I found more issues with nested types... will have to fix.

alamb commented:
>  Actually this code is a bit weird I found more issues with nested types... will have to fix.


It might make sense to get this PR in as it improves things, even if it doesn't fix everything completely. 

Also, note that @tustvold  I think is planning to work on adding upstream support in https://github.com/apache/arrow-rs/issues/4886 arrow-rs (which might automatically handle nested types), though that won't land in datafusion for a week or two

Samrose-Ahmed commented:
Oh that's great, let me make another revision and see if theirs something improving we can get in.

sarutak commented on datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:
```diff
@@ -97,15 +97,15 @@ impl<'a, R: Read> AvroArrowArrayReader<'a, R> {
-            }) => {
+            AvroSchema::Union(union_schema) => {
+                if union_schema.is_nullable() {
+                    let rec_schema = &union_schema.variants()[1];
```
This part seems incorrect. The second variants is not always non-null, but the first one can be.

sarutak replied:
Also, it's better to check if the number of variants is 2 because this logic only consider nullable types, not general union types which contains 3 or more variants.

Samrose-Ahmed commented:
I have made another revision, I had to edit some of the schema lookup logic to handle nested fields, didn't want to change the code too much.

sarutak commented on datafusion/core/Cargo.toml:
```diff
@@ -109,6 +109,8 @@ rand_distr = "0.4.3"
 regex = "1.5.4"
 rstest = "0.18.0"
 rust_decimal = { version = "1.27.0", features = ["tokio-pg"] }
+serde = "1"
```
I think `serde` is not necessary.

sarutak reviewed (commented):
This change looks good to me except one minor comment!

sarutak commented:
Oh, wait. Clippy raises some warnings.
```
error: this expression creates a reference which is immediately dereferenced by the compiler
   --> datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:115:29
    |
115 | ...                   &parent_field_name,
    |                       ^^^^^^^^^^^^^^^^^^ help: change this to: `parent_field_name`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
    = note: `-D clippy::needless-borrow` implied by `-D warnings`

error: redundant closure
   --> datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:595:38
    |
595 | ...                   .map(|v| maybe_resolve_union(v))
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `maybe_resolve_union`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
    = note: `-D clippy::redundant-closure` implied by `-D warnings`

error: could not compile `datafusion` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: the borrowed expression implements the required traits
    --> datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:1189:40
     |
1189 |           let r1 = apache_avro::to_value(&serde_json::json!({
     |  ________________________________________^
1190 | |             "headers": [
1191 | |                 {
1192 | |                     "name": "a",
...    |
1195 | |             ]
1196 | |         }))
     | |__________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
help: change this to
     |
1189 ~         let r1 = apache_avro::to_value(serde_json::json!({
1190 +             "headers": [
1191 +                 {
1192 +                     "name": "a",
1193 +                     "value": "b"
1194 +                 }
1195 +             ]
1196 ~         }))
     |

error: the borrowed expression implements the required traits
    --> datafusion/core/src/datasource/avro_to_arrow/arrow_array_reader.rs:1407:40
     |
1407 |         let r1 = apache_avro::to_value(&jv1)
     |                                        ^^^^ help: change this to: `jv1`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

error: could not compile `datafusion` (lib test) due to 4 previous errors
```

@Samrose-Ahmed  Could you fix them? Also you can check the style locally by running `ci/scripts/rust_clippy.sh` and `ci/scripts/rust_fmt.sh`.

sarutak commented:
@alamb Could you trigger GA workflows?

alamb reviewed (approved):
Thanks @Samrose-Ahmed  and @sarutak 

BTW @tustvold is working on porting / implementing this upstream: https://github.com/apache/arrow-rs/pull/4888

sarutak commented:
@alamb Thank you for letting me know. I'll check it out.

alamb commented:
In the interim, I think this PR makes DataFusion demonstrably better, and increases test coverage so I think we should merge it in. Thanks again everyone who was involved

sarutak commented:
@alamb 
BTW, will this project still open for changes related to Avro?
Actually, I'm working on a type conversion feature currently marked as `todo!()`.

alamb commented:
> BTW, will this project still open for changes related to Avro?
> Actually, I'm working on a type conversion feature currently marked as todo!().

Absolutely -- I view what @tustvold  is doing as complementary (even if somewhat duplicative). I would expect that DataFusion will continue to support avro for reading / processing, and that all end to end tests (e.g. `slt` tests) would pass without modification, even if the internals of how avro reading / processing move upstream

Does that make sense @sarutak ?

sarutak commented:
@alamb 
I understand. Thank you!
